tfx_bsl/beam/run_inference.py

```diff
@@ -95,41 +105,220 @@ class OperationType(object):
+        | '_RunInferenceCore (fixed)' >> _RunInferenceCore(
+          fixed_inference_spec_type=self._inference_spec_type,
+          catch_errors=self._with_errors)
+      )
+    elif type(self._inference_spec_type) is beam.pvalue.PCollection:
+      logging.info('RunInference on dynamic models')
+      inference_results = (
+        examples
+        | 'Join examples' >> _TemporalJoin(self._inference_spec_type)
+        | '_RunInferenceCore (dynamic)' >> _RunInferenceCore(
+          catch_errors=self._with_errors))
+    elif self._inference_spec_type is None:
+      logging.info('RunInference on queries')
+      inference_results = (
+        examples | '_RunInferenceCore (queries)' >> _RunInferenceCore(
+          catch_errors=self._with_errors))
+    else:
+      raise ValueError('Invalid type for inference_spec_type: %s'
+                       % type(self._inference_spec_type))
+
+    if self._with_errors:
+      return inference_results
+    else:
+      return inference_results['predictions']
+
+
+@beam.ptransform_fn
+@beam.typehints.with_input_types(QueryType)
+@beam.typehints.with_output_types(prediction_log_pb2.PredictionLog)
+def _RunInferenceCore(
+    queries: beam.pvalue.PCollection,
+    fixed_inference_spec_type: model_spec_pb2.InferenceSpecType = None,
+    catch_errors: bool = False
+) -> beam.pvalue.PCollection:
+  """Runs inference on queries and returns prediction logs.
+
+  This internal run inference implementation operates on queries. Internally,
+  these queries are grouped by model and inference runs in batches. If a
+  fixed_inference_spec_type is provided, this spec is used for all inference
+  requests which enables pre-configuring the model during pipeline
+  construction.
+
+  Args:
+    queries: A PCollection containing QueryType tuples
+    fixed_inference_spec_type: an optional model inference endpoint. If
+      specified, this is "preloaded" during inference and models specified in
+      query tuples are ignored. This requires the InferenceSpecType to be known
+      at pipeline creation time.
+    catch_errors: if True, runtime errors will be caught and emitted in a
+      separate 'errors' PCollection. Otherwise, runtime errors will be thrown
+      in their original location.
+
+  Returns:
+    A dict containing a prediction and error PCollection:
+    {
+      'predictions': ...,
+      'errors': ...
+    }
+    If catch_errors is False, the 'errors' PCollection will be empty.
+
+  Raises:
+    ValueError; when the fixed_inference_spec_type is invalid.
+  """
+  batched_queries = queries | 'BatchQueries' >> _BatchQueries()
+  inference_results = None
+
+  if fixed_inference_spec_type is None:
+    # operation type is determined at runtime
+    def _tag_operation_type(batch):
+      if not catch_errors:
+        return beam.pvalue.TaggedOutput(_get_operation_type(batch[0]), batch)
+      else:
+        try:
+          return beam.pvalue.TaggedOutput(_get_operation_type(batch[0]), batch)
```

Comment:
Will this throw error? It seems _get_operation_type will return str or unicode

Reply:
`_get_operation_type` has return value `Text` (str/unicode) and `beam.pvalue.TaggedOutput` accepts either str or unicode for a tag. See: https://github.com/apache/beam/blob/master/sdks/python/apache_beam/pvalue.py#L328-L343